homeassistant/components/proxmox/__init__.py

```diff
@@ -0,0 +1,148 @@
+
+REQUIREMENTS = ['proxmoxer==1.0.3']
+DOMAIN = 'proxmox'
+ENTITY_ID_FORMAT = DOMAIN + '.{}'
+UPDATE_INTERVAL = timedelta(seconds=30)
+
+_LOGGER = logging.getLogger(__name__)
+DATA_PROXMOX_NODES = 'proxmox_nodes'
+SIGNAL_PROXMOX_UPDATED = 'proxmox_updated'
+
+CONF_REALM = 'realm'
+CONF_NODES = 'nodes'
+CONF_VMS = 'vms'
+
+DEFAULT_PORT = 8006
+DEFAULT_REALM = 'pam'
+DEFAULT_VERIFY_SSL = True
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_PORT, default=DEFAULT_PORT): int,
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_REALM, default=DEFAULT_REALM): cv.string,
+        vol.Optional(CONF_VERIFY_SSL, default=DEFAULT_VERIFY_SSL): cv.boolean,
+        vol.Optional(CONF_NODES, default=[]): [cv.string],
+        vol.Optional(CONF_VMS, default=[]): [int]
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+async def async_setup(hass, config):
+    """Set up the Proxmox VE component."""
+    from proxmoxer.backends.https import AuthenticationError
+    conf = config.get(DOMAIN)
+    if conf is not None:
+        try:
+            proxmox = await get_client(conf)
+            if proxmox is None:
+                _LOGGER.exception("Error connecting to proxmox.")
+                return False
+            await setup_proxmox(hass, config)
+        except AuthenticationError:
+            # Error authenticating to Proxmox VE
+            _LOGGER.exception("Invalid username or password.")
+            return False
+    return True
+
+
+async def get_client(conf):
+    """Return Proxmox VE API client."""
+    from proxmoxer import ProxmoxAPI
+    host = conf.get(CONF_HOST)
+    port = conf.get(CONF_PORT)
+    user = conf.get(CONF_USERNAME)
+    realm = conf.get(CONF_REALM)
+    password = conf.get(CONF_PASSWORD)
+    verify_ssl = conf.get(CONF_VERIFY_SSL)
+    proxmox = ProxmoxAPI(
+        host, user=user + '@' + realm, password=password,
+        port=port, verify_ssl=verify_ssl)
+    return proxmox
+
+
+@callback
+async def setup_proxmox(hass, config):
+    """Create entities and load data from Proxmox VE."""
+    conf = config.get(DOMAIN)
+    await update_data(hass, conf)
+
+    hass.async_create_task(
+        discovery.async_load_platform(
+            hass, 'binary_sensor', DOMAIN, {}, config))
+
+    async def async_update_proxmox(now):
+        await update_data(hass, conf)
+        async_dispatcher_send(hass, SIGNAL_PROXMOX_UPDATED, None)
+
+    async_track_time_interval(hass, async_update_proxmox, UPDATE_INTERVAL)
+
+
+async def update_data(hass, conf):
+    """Update Proxmox VE data."""
+    conf_nodes = conf.get(CONF_NODES)
+    conf_vms = conf.get(CONF_VMS)
+    proxmox = await get_client(conf)
+    nodes = proxmox.nodes.get()
+    node_dict = {}
+    for node in nodes:
+        name = node['node']
+        if not bool(conf_nodes) or name in conf_nodes:
```

Comment:
Could you reverse this logic and use continue to improve readability?

Reply:
Done. Could you please review again.

Reply:
Look up guard clause, it would mean
`if <code is false>:
    continue
Real logic is cleaner and less indented
`

Reply:
I have updated the code . Could you please take a look at it